### ISSUE TITLE
META6.json needs a module identifier syntax cleanup

### ISSUE THREAD

**Issue by ghost** [perl6/toolchain-bikeshed#4]
Here's a few examples of the sort of thing S22 currently contains:

```
Foo::Bar
Foo::Bar:auth<cpan:*>:ver(1..*)
Foo::Bar:auth<cpan:baz>
Foo::Bar:ver(v1.0 .. *)
```

It's wildly inconsistent, there's barely any mention of which forms are allowed in which keys, and some of these uses of `:ver` would be a nightmare to parse in what's supposed to be a declarative file. So let's fix it.

I've got a rough set of ideas to start with:
- No part of dependency specification should _require_ a bare, unversioned module name like `Foo::Bar`. That means any part of the toolchain that currently restricts it this way is broken. To name one example, it's not possible with the current state of affairs to publish a package that provides `Foo:ver<1>` and `Foo:ver<2>` at the same time: if you're lucky only _one_ of them will fail to install.
- The various bits like `:auth<>` and `:ver<>` should be limited to non-Turing-complete syntaxes so people won't have to load an entire Perl 6 runtime to parse them, which probably means limiting to strings inside angle brackets, and simple ranges. For the sake of argument, I'd consider `:ver(v1.9.2 ^..^ v2)` simple enough.
- Foreign libraries exist and we need to deal with them. It wouldn't make sense in Perl 6 code (yet), but META6.info could have some syntax like `{"depends":["sqlite3:from<native>:ver<0.8.*>"]}` to depend on a system `libsqlite3.so.0.8.6`. Package installers would then be allowed to reconcile those deps whichever way they see fit (other packages could even list them in `provides`, if they know how to do so).

Anyway, I'll go ahead and add this to the perl6/specs meta-v1 branch later on if nobody yells "stop".

**Comment by shadowcat-mst** [comment]
Agree with all of this, basically, and I think perhaps defining a grammar for the version specs that's canonical would be a good way of ensuring people are on the same page.

My one note is that foreign library dependencies are a trifle complicated, so if possible I'd suggest we start off with an 'x-external-depends' or similar specification so people can experiment somewhat before it goes into the main spec.

**Comment by sjn** [comment]
Yeah, let's have a look at "foreign" dependencies!

We need to make it mind-numbingly trivial for packagers to figure out the correct dependencies for our software, and your suggestions are spot on for this. I especially agree that we avoid the turing-complete language mess!

Let's see if we can be thorough about this, so here are a few questions:
- Which name should we require people use when listing their "foreign" dependencies? – Different OS distributions often name their packages differently from the upstream name (e.g. "libgtk1" and "libgtk2" which both can be installed, and which have parts of the version in their names)
- Since we want to avoid turing complete metadata, how do we specify requirements that are usually checked for by a custom script? – e.g. required system resources: e.g. 1) free disk space for a database that will be set up after installation is done. e.g. 2) Availability of a specific free port number (e.g. port 7/tcp, for our App::Echo::Server application).
- How do we specify other non-library executables as dependencies? (e.g. if your module uses /usr/bin/ps on a linux system)
- How about other useful system dependencies, like a functioning /proc filesystem?

I wouldn't be surprised if there are other kinds of dependencies we could list too (I haven't really considered Windows at all here), but if we can cover the ones above, I think we'll be in a good position to make the lives of packagers a lot easier in the future. :)

**Comment by niner** [comment]
Don't depend on libgtk1 or libgtk2, depend on libgtk-3.so or even better, just "gtk-3" with $*VM.platform-library-name applied (which gives libgtk-3.so on Linux, gtk-3.dylib on OS X or gtk-3.dll on Windows). 
Of course there will still be differences between distributions on how shared libs are actually named, especially if major versions are part of the library name instead of just the extension. But I do hope that this applies to only a small part of the libraries and that those differences can be handled with very little distribution specific code.

**Comment by niner** [comment]
I think we shouldn't use "depends" for external dependencies since we don't know what syntax such dependencies could have. Depending on a foo.exe for example would already be invalid Perl 6 syntax.

**Comment by FROGGS** [comment]
@niner :+1: since that would give us the opportunity to fix #3 in one go.
